### PR TITLE
closer to browser/node compat

### PR DESCRIPTION
**Pull request by retrohacker**
WIP

### CODE REVIEW

File: lib/bridge-client/index.js [hunk 1]
```diff
@@ -465,42 +465,50 @@ BridgeClient.prototype.storeFileInBucket = function(id, token, file, opts, cb) {
...
-    });
-    file.on('error', cb);
-    file.on('end', function() {
-      ws.end();
-      // self._startDemuxing(fileSize, fileHash, id, token, cb);
-
+  // TODO: pull these cases out to functions
+  switch(file.constructor) {
+    case Readable:
+      assert(typeof fileName === 'string', 'File name must be a string.');
```
Comment (retrohacker) [inline lib/bridge-client/index.js]: should be `'opts.fileName must be a string.'`

File: lib/bridge-client/index.js [hunk 2]
```diff
@@ -465,42 +465,50 @@ BridgeClient.prototype.storeFileInBucket = function(id, token, file, opts, cb) {
...
-    file.on('data', function(data) {
-      fileSize += data.length;
-      ws.write(data);
-    });
-    file.on('error', cb);
-    file.on('end', function() {
-      ws.end();
-      // self._startDemuxing(fileSize, fileHash, id, token, cb);
-
+  // TODO: pull these cases out to functions
```
Comment (retrohacker) [inline lib/bridge-client/index.js:470]: Leaving this TODO for after the merge

File: lib/bridge-client/index.js [hunk 3]
```diff
@@ -465,42 +465,50 @@ BridgeClient.prototype.storeFileInBucket = function(id, token, file, opts, cb) {
...
+          id: id,
+          token: token
+        }, cb);
+      });
+      break;
+    case String:
+      fileName = path.basename(file).split('.crypt')[0];
+      fileSize = fs.statSync(file).size;
+      assert(typeof fileName === 'string', 'File name must be a string.');
       self._startDemuxing({
```
Comment (retrohacker) [inline lib/bridge-client/index.js:496]: we should use `setImmediate` here to break the event loop.

File: lib/bridge-client/index.js [hunk 4]
```diff
@@ -513,11 +521,9 @@ BridgeClient.prototype._startDemuxing = function(options, cb) {
     }
   );
 
-  var rs = self._store.createReadStream(fileHash);
-
   var uploadState = new UploadState({
     id: id,
-    file: rs,
+    file: file ? file : self._store.createReadStream(fileHash),
```
Comment (retrohacker) [inline lib/bridge-client/index.js]: Should we move this up into the logic of the calling function? Seems strange putting it here.

File: lib/bridge-client/index.js [hunk 5]
```diff
@@ -603,12 +604,24 @@ BridgeClient.prototype._shardUploadWorker = function(task, done) {
 BridgeClient.prototype._handleShardStream = function(options) {
   var self = this;
   const {shardStream, index, frame, uploadState, fileHash} = options;
+
+  if (uploadState.file.constructor === Readable) {
+    var tmpName = fileHash
+    var tmpFile = self._store.createWriteStream(fileHash)
+  } 
+
+  if (uploadState.file.constructor === String) {
```
Comment (retrohacker) [inline lib/bridge-client/index.js:616]: Is it possible for file to be a string at this point?

File: lib/bridge-client/index.js [hunk 6]
```diff
@@ -640,7 +654,14 @@ BridgeClient.prototype._handleShardTmpFileFinish = function(state, meta, done) {
   var self = this;
   var hash = utils.rmd160(meta.hash);
   var auditGenerator = new AuditStream(3);
-  var rs = self._store.createReadStream(meta.tmpName);
+
+  if (state.file.constructor === Readable) {
+    var rs = self._store.createReadStream(meta.tmpName);
+  } 
+
+  if (state.file.constructor === String) {
```
Comment (retrohacker) [inline lib/bridge-client/index.js:665]: Is it possible for `state.file` to be a string?

File: lib/bridge-client/index.js (hunk 4 above)
Comment (nginnever) [reply lib/bridge-client/index.js]: probably makes more sense to do a top check like the rest of the functions

File: lib/bridge-client/index.js (hunk 5 above)
Comment (nginnever) [reply lib/bridge-client/index.js:616]: file in upload state should not change and is set from the original supplied file (either string or readable stream atm).

**Pull request merged**